fix wallet loading bug on migration to new walletstore

natew opened the pull request:
This should fix issues we found this morning:

https://rainbowhaus.slack.com/archives/C090GLVBY0Y/p1750700592140369

Namely going to old version, setting up stuff, switching wallets around, then moving to new version and either:

- not selecting the right wallet
- showing wrong/loading balance

the two were related, the fix was that we had one pathway that didn't pass in the right accountAddress. now loadWallets can take that and properly sets it up.

in my testing the load was faster too now since it sets up the state instantly.

jinchung commented:
Launch in [simulator](http://localhost:29070/install/ios?virtual=https://app-team.p.rainbow.me/wallet-fix/adc27a716fef2bcc693d1e5a2111e5b45f55abdc.app.zip) or [device](http://localhost:29070/install/ios?physical=https://app-team.p.rainbow.me/wallet-fix/adc27a716fef2bcc693d1e5a2111e5b45f55abdc.ipa) for adc27a716fef2bcc693d1e5a2111e5b45f55abdc

jinchung commented:
Launch in [simulator](http://localhost:29070/install/ios?virtual=https://app-team.p.rainbow.me/wallet-fix/5c2c9fbe88a9f914a34e81ded40cb3b8fec2b6c3.app.zip) or [device](http://localhost:29070/install/ios?physical=https://app-team.p.rainbow.me/wallet-fix/5c2c9fbe88a9f914a34e81ded40cb3b8fec2b6c3.ipa) for 5c2c9fbe88a9f914a34e81ded40cb3b8fec2b6c3

maxbbb reviewed (commented):
Changes look good, but lint is failing from changes made here - https://github.com/rainbow-me/rainbow/pull/6642. Not sure if you want to fix that here or in another PR.

jinchung commented:
Launch in [simulator](http://localhost:29070/install/ios?virtual=https://app-team.p.rainbow.me/wallet-fix/f8393ef53aac8de9c54f744ab80b7f106d274e02.app.zip) or [device](http://localhost:29070/install/ios?physical=https://app-team.p.rainbow.me/wallet-fix/f8393ef53aac8de9c54f744ab80b7f106d274e02.ipa) for f8393ef53aac8de9c54f744ab80b7f106d274e02

natew commented:
@maxbbb I think my last commit to that branch didn't push before merge I was wondering what happened there. I'll fix it here.

christianbaroni commented on src/hooks/useWalletBalances.ts:
```diff
@@ -25,23 +25,30 @@ export type WalletBalanceResult = {
 const useWalletBalances = (): WalletBalanceResult => {
-  const wallets = useWallets();
-
   const nativeCurrency = userAssetsStoreManager().currency;
```
Would also move this to a `userAssetsStoreManager.getState().currency` call within the `useMemo`.

This is subscribing to the whole `userAssetsStoreManager` store, not only to `currency`. And allowing `currency` to update here before the wallets summary data when `currency` changes will cause e.g., for a USD → EUR currency change, USD values to briefly render as EUR values given the logic in the `useMemo`.

jinchung commented:
Launch in [simulator](http://localhost:29070/install/ios?virtual=https://app-team.p.rainbow.me/wallet-fix/a7d6432db9fd058a35fe4626d1fca9a7a57a6ac2.app.zip) or [device](http://localhost:29070/install/ios?physical=https://app-team.p.rainbow.me/wallet-fix/a7d6432db9fd058a35fe4626d1fca9a7a57a6ac2.ipa) for a7d6432db9fd058a35fe4626d1fca9a7a57a6ac2

jinchung commented:
Launch in [simulator](http://localhost:29070/install/ios?virtual=https://app-team.p.rainbow.me/wallet-fix/d1ef990aaa40fba55a8b0f26af0b843ca23a76e3.app.zip) or [device](http://localhost:29070/install/ios?physical=https://app-team.p.rainbow.me/wallet-fix/d1ef990aaa40fba55a8b0f26af0b843ca23a76e3.ipa) for d1ef990aaa40fba55a8b0f26af0b843ca23a76e3

jinchung commented:
Launch in [simulator](http://localhost:29070/install/ios?virtual=https://app-team.p.rainbow.me/wallet-fix/d49d0b6d0a3809d2c8cf77395a983f89fabe4e1f.app.zip) or [device](http://localhost:29070/install/ios?physical=https://app-team.p.rainbow.me/wallet-fix/d49d0b6d0a3809d2c8cf77395a983f89fabe4e1f.ipa) for d49d0b6d0a3809d2c8cf77395a983f89fabe4e1f

natew commented on src/state/internal/utils/debugStoreUtils.ts:
done

natew commented on src/components/coin-row/ListCoinRow.js:
```diff
@@ -1,13 +1,13 @@
-import React, { Fragment, useCallback } from 'react';
+import React, { Fragment, useCallback, useMemo } from 'react';
```
this useMemo was entirely missing my type checker noticed it but not sure how it was working before or if this file is unused

natew commented on src/state/wallets/walletsStore.ts:
```diff
@@ -546,18 +552,14 @@ async function refreshWalletsInfo({ wallets, useCachedENS }: GetENSInfoProps) {
+async function refreshAccountInfo(accountIn: RainbowAccount, useCachedENS = false): Promise<RainbowAccount> {
+  const account = {
+    ...accountIn,
+    label: removeFirstEmojiFromString(accountIn.label || addressAbbreviation(accountIn.address, 4, 4)),
```
this fixes emoji showing on label by ensuring it sets the cleaned label

christianbaroni reviewed (approved):
This could be addressed in a follow-up PR since it's more so a performance optimization at this point, but my only remaining reservations have to do with this change, and with `loadWallets` in general. Would like to better understand specifically what issue this addresses:

https://github.com/rainbow-me/rainbow/pull/6705/files#diff-32eb751238a0f34664a9f95644b355957491fdb36dcbaad4897394c6ec950b96R126

My sense is that most of the work `loadWallets` is doing serves no purpose, or is at least avoidable. With the exception of the metadata refreshing, isn't `loadWallets` implicitly assuming the store's selected wallet state is incorrect? (Or is missing due to not yet being initialized — but that should only occur once and could be handled explicitly as part of wallet creation, plus a migration for existing users.)

In the majority of cases, it seems like we're reading from the keychain, then setting what's likely already set in the wallets store (the selected wallet/address), and then setting it in the keychain *again* via the `setSelectedWallet` call in `loadWallets` that gets triggered by passing `walletAddress` to `loadWallets`.

Granted, this now only occurs once on launch, and it's likely worth guarding against selected wallet mismatches until we have stronger guarantees, but I think at the very least we could skip some of this work if we can determine that there's not a mismatch between the keychain and the store that needs correcting.

Also, regarding `initializeWallet` and the fact that it awaits `loadWallets` — it seems unnecessary to make wallet initialization await the wallet metadata refresh.

Otherwise, rest of the changes look good to me.

jinchung commented:
Launch in [simulator](http://localhost:29070/install/ios?virtual=https://app-team.p.rainbow.me/wallet-fix/1631929cdfefc113967aa18d1eefc69084d42522.app.zip) or [device](http://localhost:29070/install/ios?physical=https://app-team.p.rainbow.me/wallet-fix/1631929cdfefc113967aa18d1eefc69084d42522.ipa) for 1631929cdfefc113967aa18d1eefc69084d42522

natew commented:
@christianbaroni went back and did the back and forth a few more times with and without the logic and it seems it does work without it, though I had at one point yesterday seen the bug happen without it and then fix with it, which leaves me wondering what's going on.

Either way after re-reading over everything I agree that it seems unnecessary. I removed it and fixed one more piece of logic (checking if it's just "0x" to determine if it's actually not set yet) and that maybe is the better fix anyway.
